subsys/fs/fuse_fs_access.c

```diff
@@ -0,0 +1,530 @@
+	} else {
+		err = fs_opendir(&dir, path);
+	}
+
+	if (err) {
+		return -ENOEXEC;
+	}
+
+	stat.st_dev = 0;
+	stat.st_ino = 0;
+	stat.st_nlink = 0;
+	stat.st_uid = getuid();
+	stat.st_gid = getgid();
+	stat.st_rdev = 0;
+	stat.st_atime = 0;
+	stat.st_mtime = 0;
+	stat.st_ctime = 0;
+	stat.st_mode = S_IFDIR | S_IRWX_DIR;
+	stat.st_size = 0;
+	stat.st_blksize = 0;
+	stat.st_blocks = 0;
+
+	filler(buf, ".", &stat, 0);
+	filler(buf, "..", &stat, 0);
+
+	do {
+		err = fs_readdir(&dir, &entry);
+		if (err) {
+			break;
+		}
+
+		if (entry.name[0] == DIR_END) {
+			break;
+		}
+
+		if (entry.type == FS_DIR_ENTRY_DIR) {
+			stat.st_mode = S_IFDIR | S_IRWX_DIR;
+			stat.st_size = 0;
+		} else {
+			stat.st_mode = S_IFREG | S_IRW_FILE;
+			stat.st_size = entry.size;
+		}
+
+		if (filler(buf, entry.name, &stat, 0)) {
+			break;
+		}
+
+	} while (1);
+
+	fs_closedir(&dir);
+
+	return err;
+
+}
+
+static int fuse_fs_access_create(const char *path, mode_t mode,
+			     struct fuse_file_info *fi)
+{
+	int err;
+	size_t handle;
+
+	ARG_UNUSED(mode);
+
+	if (is_mount_point(path)) {
+		return -ENOENT;
+	}
+
+	handle = get_new_file_handle();
+	if (handle == INVALID_FILE_HANDLE) {
+		return -ENOMEM;
+	}
+
+	fi->fh = handle;
+
+	err = fs_open(&files[handle], path);
+	if (err != 0) {
+		release_file_handle(handle);
+		fi->fh = INVALID_FILE_HANDLE;
+		return err;
+	}
+
+	return 0;
+}
+
+static int fuse_fs_access_open(const char *path, struct fuse_file_info *fi)
+{
+	return fuse_fs_access_create(path, 0, fi);
+}
+
+static int fuse_fs_access_release(const char *path, struct fuse_file_info *fi)
+{
+	ARG_UNUSED(path);
+
+	if (fi->fh == INVALID_FILE_HANDLE) {
+		return -EINVAL;
+	}
+
+	fs_close(&files[fi->fh]);
+
+	release_file_handle(fi->fh);
+
+	return 0;
+}
+
+static int fuse_fs_access_read(const char *path, char *buf, size_t size,
+		off_t off, struct fuse_file_info *fi)
+{
+	int err;
+
+	ARG_UNUSED(path);
+
+	if (fi->fh == INVALID_FILE_HANDLE) {
+		return -EINVAL;
+	}
+
+	err = fs_seek(&files[fi->fh], off, FS_SEEK_SET);
+	if (err != 0) {
+		return err;
+	}
+
+	err = fs_read(&files[fi->fh], buf, size);
+
+	return err;
+}
+
+static int fuse_fs_access_write(const char *path, const char *buf, size_t size,
+		off_t off, struct fuse_file_info *fi)
+{
+	int err;
+
+	ARG_UNUSED(path);
+
+	if (fi->fh == INVALID_FILE_HANDLE) {
+		return -EINVAL;
+	}
+
+	err = fs_seek(&files[fi->fh], off, FS_SEEK_SET);
+	if (err != 0) {
+		return err;
+	}
+
+	err = fs_write(&files[fi->fh], buf, size);
+
+	return err;
+}
+
+static int fuse_fs_access_ftruncate(const char *path, off_t size,
+				struct fuse_file_info *fi)
+{
+	int err;
+
+	ARG_UNUSED(path);
+
+	if (fi->fh == INVALID_FILE_HANDLE) {
+		return -EINVAL;
+	}
+
+	err = fs_truncate(&files[fi->fh], size);
+
+	return err;
+}
+
+static int fuse_fs_access_truncate(const char *path, off_t size)
+{
+	int err;
+	static struct fs_file_t file;
+
+	err = fs_open(&file, path);
+	if (err != 0) {
+		return err;
+	}
+
+	err = fs_truncate(&file, size);
```

Comment:
Doses NFFS supports truncate at all? I thought it doesn't.

Reply:
NFFS does not support truncate but correctly returns -ENOTSUP so FUSE will return this correctly to the host user space program. Note that the FAT file system implementation in zephyr does support truncate.